source/common/http/filter/fault_filter.cc

```diff
@@ -69,7 +80,11 @@ FaultFilter::~FaultFilter() { ASSERT(!delay_timer_); }
 // if we inject a delay, then we will inject the abort in the delay timer
 // callback.
 FilterHeadersStatus FaultFilter::decodeHeaders(HeaderMap& headers, bool) {
-  if (!matchesTargetCluster()) {
+  if (!matchesTargetUpstreamCluster()) {
+    return FilterHeadersStatus::Continue;
+  }
+
+  if (!matchesDownstreamNodes(headers)) {
     return FilterHeadersStatus::Continue;
```

Comment:
This is what I was worried about. We are doing two header match passes. We could fold this into a single header match, can't we? It makes the configuration simple as well.

With this PR, I can do match based on downstreams either using the header or using the downstream_nodes option. Instead of supporting multiple ways for doing a single thing, I would suggest providing a single well defined way of achieving X, whatever it is.